rust/kernel/spi.rs

```diff
@@ -0,0 +1,222 @@
+            $($method: true),+,
+            ..$crate::spi::USE_NONE
+        };
+    };
+}
+
+impl DriverRegistration {
+    fn new(this_module: &'static crate::ThisModule, name: &'static CStr) -> Self {
+        DriverRegistration {
+            this_module,
+            name,
+            registered: false,
+            spi_driver: bindings::spi_driver::default(),
+        }
+    }
+
+    // FIXME: Add documentation
+    pub fn new_pinned<T: SpiMethods>(
+        this_module: &'static crate::ThisModule,
+        name: &'static CStr,
+    ) -> Result<Pin<Box<Self>>> {
+        let mut registration = Pin::from(Box::try_new(Self::new(this_module, name))?);
+
+        registration.as_mut().register::<T>()?;
+
+        Ok(registration)
+    }
+
+    unsafe extern "C" fn probe_wrapper<T: SpiMethods>(
+        spi_dev: *mut bindings::spi_device,
+    ) -> c_types::c_int {
+        // SAFETY: The spi_dev pointer is provided by the kernel and is sure to be valid
+        match T::probe(SpiDevice::from_ptr(spi_dev)) {
+            Ok(_) => 0,
+            Err(e) => e.to_kernel_errno(),
+        }
+    }
+
+    unsafe extern "C" fn remove_wrapper<T: SpiMethods>(
+        spi_dev: *mut bindings::spi_device,
+    ) -> c_types::c_int {
+        // SAFETY: The spi_dev pointer is provided by the kernel and is sure to be valid
+        match T::remove(SpiDevice::from_ptr(spi_dev)) {
+            Ok(_) => 0,
+            Err(e) => e.to_kernel_errno(),
+        }
+    }
+
+    unsafe extern "C" fn shutdown_wrapper<T: SpiMethods>(spi_dev: *mut bindings::spi_device) {
+        // SAFETY: The spi_dev pointer is provided by the kernel and is sure to be valid
+        T::shutdown(SpiDevice::from_ptr(spi_dev))
+    }
+
+    // FIXME: Add documentation
+    pub fn register<T: SpiMethods>(self: Pin<&mut Self>) -> Result {
+        fn maybe_get_wrapper<F>(vtable_value: bool, func: F) -> Option<F> {
+            match vtable_value {
+                false => None,
+                true => Some(func),
+            }
+        }
+
+        let this = unsafe { self.get_unchecked_mut() };
+        if this.registered {
+            return Err(Error::EINVAL);
+        }
+
+        this.spi_driver.driver.name = this.name.as_ptr() as *const c_types::c_char;
+        this.spi_driver.probe =
+            maybe_get_wrapper(T::TO_USE.probe, DriverRegistration::probe_wrapper::<T>);
+        this.spi_driver.remove =
+            maybe_get_wrapper(T::TO_USE.remove, DriverRegistration::remove_wrapper::<T>);
+        this.spi_driver.shutdown = maybe_get_wrapper(
+            T::TO_USE.shutdown,
+            DriverRegistration::shutdown_wrapper::<T>,
+        );
+
+        let res =
+            unsafe { bindings::__spi_register_driver(this.this_module.0, &mut this.spi_driver) };
+
+        match res {
+            0 => {
+                this.registered = true;
+                Ok(())
+            }
+            _ => Err(Error::from_kernel_errno(res)),
```

Comment:
I think it's better just to have a `if` for the error case, and the fallthrough to the `Ok` case.

Reply:
Indeed, please always try to keep the happy path non-indented.